flink-end-to-end-tests/test-scripts/common.sh

```diff
@@ -799,3 +799,32 @@ function extract_job_id_from_job_submission_return() {
     echo "$JOB_ID"
 }
 
+
+#
+# NOTE: This function requires at least Bash version >= 4. Mac OS in 2020 still ships 3.x
+#
+
+run_test_with_timeout() {
```

Comment:
Consider this script:

```
#!/usr/bin/env bash


f1() {
	local v1="value"
	echo $v1
	f2() {
		echo "f2 $v1"
	}
	(f2)
}

f1
f2
```

When running it, it will output (bash 5.0.17):
```
value
f2 value
f2 
```

We can see that `local` variables are visible in subshells. Therefore, we can declare `TEST_TIMEOUT_SECONDS` and  `TEST_COMMAND` as `local`. 

Furthermore, `f2` will be accessible after running `f1`, i.e., after running `f1`, we have defined a regular function that is accessible from outside of `f1`. I think that's acceptable. However, there is a trick to limit the scope of a nested function using subshells:
```
f1() ( 
	local v1="value"
	echo $v1
	f2() {
		echo "f2 $v1"
	}
	(f2)
)
```
Note that the regular parentheses (not the curly ones).

Output:

```
value
f2 value
./test.sh: line 15: f2: command not found
```

Reply:
Thank you for your detailed answer. I have played with this (locally ;-) ) and decided to move `kill_test_watchdog` outside of `run_test_with_timeout`. I'm registering the function with the `on_exit` callback, thus it needs to be accessible outside of run_test_with_timeout`. 
I didn't spend much thought on this "function inside function". I just thought it might be nicer to indicate where the inner function belongs to.

Reply:
> I'm registering the function with the on_exit callback, thus it needs to be accessible outside of run_test_with_timeout.

Sorry, missed that.

Reply:
`local` can still be added I think. The variables are only used in the subshell.

Reply:
Yes, I will merge it with `local`.